ansible_base/oauth2_provider/models/access_token.py

```diff
@@ -0,0 +1,93 @@
+import oauth2_provider.models as oauth2_models
+from django.conf import settings
+from django.db import connection, models
+from django.utils.timezone import now
+from django.utils.translation import gettext_lazy as _
+from oauthlib import oauth2
+
+from ansible_base.lib.abstract_models.common import CommonModel
+from ansible_base.lib.utils.models import prevent_search
+from ansible_base.lib.utils.settings import get_setting
+from ansible_base.oauth2_provider.utils import is_external_account
+
+activitystream = object
+if 'ansible_base.activitystream' in settings.INSTALLED_APPS:
+    from ansible_base.activitystream.models import AuditableModel
+
+    activitystream = AuditableModel
+
+
+class OAuth2AccessToken(CommonModel, oauth2_models.AbstractAccessToken, activitystream):
+    router_basename = 'token'
+    ignore_relations = ['refresh_token']
+
+    class Meta(oauth2_models.AbstractAccessToken.Meta):
+        verbose_name = _('access token')
+        ordering = ('id',)
+        swappable = "OAUTH2_PROVIDER_ACCESS_TOKEN_MODEL"
+
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
+        on_delete=models.CASCADE,
+        blank=True,
+        null=True,
+        related_name="access_tokens",
+        help_text=_('The user representing the token owner'),
+    )
+    # Overriding to set related_name
+    application = models.ForeignKey(
+        settings.OAUTH2_PROVIDER_APPLICATION_MODEL,
+        on_delete=models.CASCADE,
+        blank=True,
+        null=True,
+        related_name='access_tokens',
+    )
+    description = models.TextField(
+        default='',
+        blank=True,
+    )
+    last_used = models.DateTimeField(
+        null=True,
+        default=None,
+        editable=False,
+    )
+    scope = models.CharField(
+        blank=True,
+        default='write',
+        max_length=32,
+        choices=[('read', 'read'), ('write', 'write')],
+        help_text=_("Allowed scopes, further restricts user's permissions. Must be a simple space-separated string with allowed scopes ['read', 'write']."),
+    )
+    token = prevent_search(
+        models.CharField(
+            max_length=255,
+            unique=True,
+        )
+    )
+    updated = None  # Tracked in CommonModel with 'modified', no need for this
+
+    def is_valid(self, scopes=None):
+        valid = super(OAuth2AccessToken, self).is_valid(scopes)
+        if valid:
+            self.last_used = now()
+
+            def _update_last_used():
+                if OAuth2AccessToken.objects.filter(pk=self.pk).exists():
+                    self.save(update_fields=['last_used'])
+
+            connection.on_commit(_update_last_used)
```

Comment:
This is in `is_valid()` which is what DOT uses to check if the token is expired or not, so this seems like a reasonable place for this (though it could probably stand to be commented what's going on).

Basically DOT calls this method to see if it's still a valid token, then we say "if it is valid -- on the next commit bump the `last_used` because it was accessed".